### PR TITLE
Schema: Split metadata table out into separate collection based tables

### PR DESCRIPTION
**Pull request by mapkyca**
## Here's what I fixed or added:

Splits metadata tables out into their own tables.

## Here's why I did it:

This is the onward goal of improving the database schema bit by bit.

Splitting metadata out into their own tables should make things a bit faster. 

More importantly makes the schema easier to work with, for example, we can offload data integrity (e.g. deleting metadata) off to the database layer to manage. 

It also means we might be able to move over utf8mb4 at some point in the future (#2451), for which the length of indexes proved blockers.

## Checklist: (`[x]` to check/tick the boxes)

- [x] This pull request addresses a single issue
- [ ] If this code includes interface changes, I've included screenshots in this Pull Request thread
- [x] I've adhered to [Known's style guide](http://docs.withknown.com/en/latest/developers/standards/) ([these codesniffer rules](http://docs.withknown.com/en/latest/developers/testing/#code-style-testing) might help!)
- [x] My git branch is named in a descriptive way - i.e., yourname-summary-of-issue
- [x] I've tested my code in-browser
- [ ] My code contains descriptive comments
- [ ] I've added tests where applicable, and...
- [x] I can run the [unit tests](http://docs.withknown.com/en/latest/developers/testing/#unit-testing) successfully.

### CODE REVIEW

**Comment by Lewiscowles1986** [comment]
This isn't working when rebased against an existing deploy using postgres for heroku. Specifically it does not find the sole existing user, which it's been confirmed at the database level still exists.

**Comment by Lewiscowles1986** [comment]
These two fixed being able to see my site, but not logging in

```
INSERT INTO config_metadata
SELECT _id, name, value FROM metadata WHERE collection = 'config'
ON CONFLICT DO NOTHING;
```

```
INSERT INTO entities_metadata
SELECT _id, name, value FROM metadata WHERE collection = 'entities'
ON CONFLICT DO NOTHING;
```

there is an interesting property in that the name and _id do not seem to be unique constraint enforced

**Comment by mapkyca** [comment]
Yeah, it would actually help if I _executed_ the patch script. Serves me for doing late night coding. Try now.

(obviously back up first)

**Comment by mapkyca** [comment]
Regarding unique. Yes, this is known and it is Known. Saving an entity clears all metadata for that entity and resets. Not the _best_ way, just _a_ way.

Handling metadata more efficiently is definitely on the list, but I'm not there yet.

**Comment by Lewiscowles1986** [comment]
```sql
TRUNCATE TABLE config_metadata;
ALTER TABLE config_metadata ADD UNIQUE(_id, name);

INSERT INTO config_metadata
SELECT _id, name, value FROM metadata WHERE collection = 'config'
ON CONFLICT DO NOTHING;

TRUNCATE TABLE entities_metadata;
ALTER TABLE entities_metadata ADD UNIQUE(_id, name);

INSERT INTO entities_metadata
SELECT _id, name, value FROM metadata WHERE collection = 'entities'
ON CONFLICT DO NOTHING;
```

This is my just-now done fix. It does mean I'll need to run some checks to ensure it's not run per-deploy.

File: Idno/Data/MySQL.php [hunk 1]
```diff
@@ -75,6 +75,7 @@ function checkAndUpgradeSchema()
                             2017032001,
                             2019060501,
                             2019121401,
+                            2020042101,
```
Comment (Lewiscowles1986) [inline Idno/Data/MySQL.php:78]: is this one supposed to update too?

Comment (mapkyca) [reply Idno/Data/MySQL.php:78]: Only if you're running with the mysql engine. 

Scripts only execute for the engine you're on.

File: warmup/schemas/mysql/mysql.sql [hunk 2]
```diff
@@ -29,6 +29,16 @@ CREATE TABLE IF NOT EXISTS `config_search` (
...
 
+CREATE TABLE IF NOT EXISTS `config_metadata` (
+  `_id` varchar(32) NOT NULL,
+  `name` varchar(64) NOT NULL,
+  `value` text NOT NULL,
+  KEY `value` (`value`(255)),
+  KEY `name` (`name`),
+  KEY (`_id`),
+  CONSTRAINT `cm_id_id` FOREIGN KEY (`_id`) REFERENCES `config` (`_id`) ON DELETE CASCADE ON UPDATE CASCADE
+) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```
Comment (Lewiscowles1986) [inline warmup/schemas/mysql/mysql.sql:40]: UNIQUE constraint could go here. Basically in a fairly new (April 1st) Install, I had duplicate values stored which were from the application when running the `INSERT INTO SELECT FROM WHERE` part of the queries

File: warmup/schemas/postgres/2020042101.sql [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+INSERT INTO config_metadata SELECT _id, name, value FROM metadata WHERE collection = 'config';
+
+
+
+CREATE TABLE IF NOT EXISTS reader_metadata (
+  _id varchar(32) NOT NULL,
+  name varchar(64) NOT NULL,
+  value text NOT NULL,
+  FOREIGN KEY (_id) REFERENCES reader (_id) ON DELETE CASCADE ON UPDATE CASCADE
+);
```
Comment (Lewiscowles1986) [inline warmup/schemas/postgres/2020042101.sql:37]: UNIQUE constraint could go here. Basically in a fairly new (April 1st) Install, I had duplicate values stored which were from the application when running the INSERT INTO SELECT FROM WHERE part of the queries

File: warmup/schemas/postgres/postgres.sql [hunk 4]
```diff
@@ -28,6 +28,17 @@ CREATE TABLE IF NOT EXISTS config_search (
   FOREIGN KEY (_id) REFERENCES config (_id) ON DELETE CASCADE ON UPDATE CASCADE
 );
 
+CREATE TABLE IF NOT EXISTS config_metadata (
+  _id varchar(32) NOT NULL,
+  name varchar(64) NOT NULL,
+  value text NOT NULL,
+  FOREIGN KEY (_id) REFERENCES config (_id) ON DELETE CASCADE ON UPDATE CASCADE
```
Comment (Lewiscowles1986) [inline warmup/schemas/postgres/postgres.sql:35]: I think both warmup and patch need the unique constraint added. It is quite confusing both exist to be honest

File: warmup/schemas/postgres/2020042101.sql (hunk 3 above)
Comment (mapkyca) [reply warmup/schemas/postgres/2020042101.sql:37]: Have to check how this affects the storage of array values.

Can't remember whether arrays are serialised in metadata or stored multiple times (I think the latter). Switched between projects so many times I can't remember off the top of my head

File: warmup/schemas/postgres/postgres.sql (hunk 4 above)
Comment (mapkyca) [reply warmup/schemas/postgres/postgres.sql:35]: Warmup is canonical, patch is to roll up existing users as defined by the schema version number

File: Idno/Data/MySQL.php (hunk 1 above)
Comment (Lewiscowles1986) [reply Idno/Data/MySQL.php:78]: Right. Well the postgres edits are not in-line with the MySQL ones then. It's one of the parts that was confusing.

File: warmup/schemas/postgres/postgres.sql (hunk 4 above)
Comment (Lewiscowles1986) [reply warmup/schemas/postgres/postgres.sql:35]: so it does protect against a patch running a second time?

Comment (mapkyca) [reply warmup/schemas/postgres/postgres.sql:35]: In a crude way. Patch files update the schema version as they go.

Something I want to do, but probably will never have time to do, is move this stuff all over to doctrine dbal and have schema definitions stored in json so they can be managed automatically.

Done this for another project and it removes a lot of headaches.

File: Idno/Data/MySQL.php (hunk 1 above)
Comment (mapkyca) [reply Idno/Data/MySQL.php:78]: Well, some patches are only relevant to specific database engines hence why the difference (also, Postgres isn't so much a well trodden path)

**Pull request merged**